### ISSUE TITLE
[question] Should the win_bash argument of run() be "True" when executed from MSYS2 bash

### ISSUE THREAD

**Issue by wrightsg** [conan-io/conan#5071]
To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

Conan version: 1.14.1
OS: Windows 10, 64-bit
Environment: Running ``conan`` from ``MSYS2`` ``bash`` shell (more specifically the ``MINGW64`` environment, i.e. for compiling native Windows 64-bit applications); ``conan`` was installed using the ``mingw64/mingw-w64-x86_64-python3-pip`` package; furthermore I'm using ``gcc`` 8.3.0

I've ran into some problem while installing the [OpenSSL](https://github.com/conan-community/conan-openssl) and [libjpeg](https://github.com/bincrafters/conan-libjpeg) packages.

For example, running

```
conan install --build missing -s compiler=gcc -s compiler.libcxx=libstdc++11 -s compiler.version=8 libjpeg/9c@bincrafters/stable
```
will result in the following output

```
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++11
compiler.version=8
os=Windows
os_build=Windows
[options]
[build_requires]
[env]

Installing package: libjpeg/9c@bincrafters/stable
Requirements
    libjpeg/9c@bincrafters/stable from 'conan-center' - Cache
Packages
    libjpeg/9c@bincrafters/stable:9965605309592d7e617ec929633249a2031e4263 - Build

Downloading conan_sources.tgz
[==================================================] 5.5KB/5.5KB
Decompressing conan_sources.tgz: 100%|| 5.35k/5.35k [00:00<00:00, 609kB/s]s]
libjpeg/9c@bincrafters/stable: Configuring sources in C:/msys64/home/swright/.conan/data/libjpeg/9c/bincrafters/stable/source
[==================================================] 1.0MB/1.0MB
libjpeg/9c@bincrafters/stable: Copying sources to build folder
libjpeg/9c@bincrafters/stable: Building your package in C:/msys64/home/swright/.conan/data/libjpeg/9c/bincrafters/stable/build/9965605309592d7e617ec929633249a2031e4263
libjpeg/9c@bincrafters/stable: Generator txt created conanbuildinfo.txt
libjpeg/9c@bincrafters/stable: Calling build()
libjpeg/9c@bincrafters/stable: WARN: Error running `configure --help`: Error 1 while executing /c/msys64/home/swright/.conan/data/libjpeg/9c/bincrafters/stable/build/9965605309592d7e617ec929633249a2031e4263/source_subfolder/configure --help
libjpeg/9c@bincrafters/stable: Calling:
 > /c/msys64/home/swright/.conan/data/libjpeg/9c/bincrafters/stable/build/9965605309592d7e617ec929633249a2031e4263/source_subfolder/configure --enable-shared=no --enable-static=yes --prefix=/c/msys64/home/swright/.conan/data/libjpeg/9c/bincrafters/stable/package/9965605309592d7e617ec929633249a2031e4263 --build=x86_64-w64-mingw32 --host=x86_64-w64-mingw32
libjpeg/9c@bincrafters/stable:
libjpeg/9c@bincrafters/stable: ERROR: Package '9965605309592d7e617ec929633249a2031e4263' build failed
libjpeg/9c@bincrafters/stable: WARN: Build folder C:/msys64/home/swright/.conan/data/libjpeg/9c/bincrafters/stable/build/9965605309592d7e617ec929633249a2031e4263
ERROR: libjpeg/9c@bincrafters/stable: Error in build() method, line 85
        self._build_configure()
while calling '_build_configure', line 77
        env_build.configure(configure_dir=self._source_subfolder, args=config_args)
        ConanException: Cannot recognize the Windows subsystem, install MSYS2/cygwin or specify a build_require to apply it.
```

After some research I came to the conclusion that this error is somehow related to the ``win_bash`` parameter of the ``run()`` function. Both packages mentioned above seem to set ``win_bash`` to ``True`` when ``conan install`` is run from my environment.

To debug the issue I created a dummy package (i.e. ``conan new mypkg/0.1 -t``) and modified the ``build()`` function as follows

```
def build(self):
    self.run("cmake --version", win_bash=True)

    cmake = CMake(self)
    cmake.configure(source_folder="hello")
    cmake.build()
```

Running

```
conan create -s compiler=gcc -s compiler.libcxx=libstdc++11 -s compiler.version=8 . mypkg/testing
```

with this dummy package results in the same error. However, package creation does succeed if ``win_bash`` is set to ``False``.

Summary: for some reason, ``conan`` "cannot recognize the Windows subsystem" when executed from ``MSYS2`` ``bash`` shell and ``win_bash`` is set to ``True``. My question is if ``win_bash`` is intended to be set to ``False`` in such an environment or if I'm missing something else.

**Comment by memsharded** [comment]
Hi @wrightsg 

Thanks for your detailed report, and your reproducible case.

@SSE4 has done some work in the MSYS and MinGW ecosystems, IIRC. If that is the case, could you please have a look at this issue? Many thanks!

**Comment by SSE4** [comment]
@wrightsg could you execute `uname -a` in your shell and lemme know what does it output?
I am pretty sure `detect_windows_subsystem` was able to recognize MSYS2, MinGW, Cygwin shells, but there are some variations of it, and probably we have missed some variant

**Comment by SSE4** [comment]
I personally have the following:
MSYS shell: `MSYS_NT-10.0 DESKTOP-E3NFU1V 3.0.6(0.338/5/3) 2019-04-15 06:44 x86_64 Msys`
MinGW32 shell: `MINGW32_NT-10.0 DESKTOP-E3NFU1V 3.0.6(0.338/5/3) 2019-04-15 06:44 x86_64 Msys`
MinGW64 shell: `MINGW64_NT-10.0 DESKTOP-E3NFU1V 3.0.6(0.338/5/3) 2019-04-15 06:44 x86_64 Msys`
looking at the code, `detect_windows_subsystem` recognizes all of them
I am trying to reproduce it now...

**Comment by SSE4** [comment]
yeah, reproduced:
```
ERROR: mypkg/0.1@mypkg/testing: Error in build() method, line 29
        self.run("cmake --version", win_bash=True)
        ConanException: Command only for Windows operating system
```

**Comment by SSE4** [comment]
so, the issue is that currently it excepts only versions 1.x and 2.x, by MSYS has version 3.0.6
https://github.com/conan-io/conan/blob/develop/conans/client/tools/oss.py#L327

**Comment by wrightsg** [comment]
Thanks for investigating. Indeed I'm also running version 3.0.6:

``MINGW64_NT-10.0-16299 londonmule 3.0.6-338.x86_64 2019-04-15 06:44 UTC x86_64 Msys``

Is there any way to workaround this issue with ``conan`` 1.14.4?

**Comment by SSE4** [comment]
I think only if you find a way to override uname or just use older msys...

**Comment by SSE4** [comment]
or if you can just use `win_bash=False`

**Comment by memsharded** [comment]
A fix has been done in #5078, will be released in next Conan 1.15.
Many thanks @SSE4 and @wrightsg for the feedback!
